File: vscode-extension/src/antares/core/cwe.ts
// Port of antares_cli/core/cwe.py — deterministic CWE id normalization/validation.

import { CweDatabase } from "../knowledge/cweDatabase";

const CWE_STRICT_PATTERN = /^(?:CWE[-_\s]*)?0*(\d+)$/i;
const CWE_SEARCH_PATTERN = /CWE-(\d+)/i;

export class CweIdError extends Error {}

// Normalize a CWE identifier to canonical "CWE-NNN" form.
// strict=true: input must be a standalone reference ("22", "cwe-22", "CWE_22").
// strict=false: extract the first CWE-NNN anywhere, fall back to bare digits,
//   else return the original string unchanged.
export function normalizeCweId(rawCweId: string, strict = true): string {
  const text = rawCweId.trim();

  if (strict) {
    const match = text.match(CWE_STRICT_PATTERN);
    if (match === null) {
      throw new CweIdError(`Invalid CWE ID: ${JSON.stringify(rawCweId)}`);
    }
    const numericId = Number.parseInt(match[1], 10);
    if (numericId <= 0) {
      throw new CweIdError(`Invalid CWE ID: ${JSON.stringify(rawCweId)}`);
    }
    return `CWE-${numericId}`;
  }

  const searchMatch = text.match(CWE_SEARCH_PATTERN);
  if (searchMatch) {
    return `CWE-${Number.parseInt(searchMatch[1], 10)}`;
  }
  if (/^\d+$/.test(text)) {
    return `CWE-${Number.parseInt(text, 10)}`;
  }
  return text;
}

// Normalize, validate against the catalog, and de-duplicate while preserving order.
export function normalizeCweIds(rawCweIds: string[], cweDatabase: CweDatabase): string[] {
  const normalizedIds: string[] = [];
  const seen = new Set<string>();
  for (const rawCweId of rawCweIds) {
    const normalizedId = normalizeCweId(rawCweId);
    if (cweDatabase.getById(normalizedId) === undefined) {
      throw new CweIdError(`Unknown CWE ID: ${normalizedId}`);
    }
    if (!seen.has(normalizedId)) {
      normalizedIds.push(normalizedId);
      seen.add(normalizedId);
    }
  }
  return normalizedIds;
}

// Parse a comma-separated CLI CWE list into raw id tokens.
export function parseCweIdList(rawValue: string | null | undefined): string[] {
  if (rawValue === null || rawValue === undefined) {
    return [];
  }
  return rawValue
    .split(",")
    .map((part) => part.trim())
    .filter((part) => part.length > 0);
}

File: src/antares_cli/core/cwe.py
"""Deterministic CWE ID normalization and validation."""

from __future__ import annotations

import re

from antares_cli.knowledge.cwe_database import CweDatabase

_CWE_STRICT_PATTERN = re.compile(r"^(?:CWE[-_\s]*)?0*(\d+)$", re.IGNORECASE)
_CWE_SEARCH_PATTERN = re.compile(r"CWE-(\d+)", re.IGNORECASE)


class CweIdError(ValueError):
    """Raised when a CWE ID cannot be normalized or validated."""


def normalize_cwe_id(raw_cwe_id: str, *, strict: bool = True) -> str:
    """Normalize a CWE identifier to the canonical ``CWE-NNN`` form.

    strict=True (default): input must be a standalone CWE reference like
        "22", "cwe-22", "CWE_22". Raises CweIdError if unparseable.

    strict=False: extracts the first CWE-NNN from anywhere in the string
        (e.g. "CWE-798: Hard-coded Credentials" → "CWE-798").
        Falls back to bare-digit handling ("22" → "CWE-22").
        Returns the original string if nothing matches.
    """
    text = raw_cwe_id.strip()

    if strict:
        match = _CWE_STRICT_PATTERN.match(text)
        if match is None:
            raise CweIdError(f"Invalid CWE ID: {raw_cwe_id!r}")
        numeric_id = int(match.group(1))
        if numeric_id <= 0:
            raise CweIdError(f"Invalid CWE ID: {raw_cwe_id!r}")
        return f"CWE-{numeric_id}"

    search_match = _CWE_SEARCH_PATTERN.search(text)
    if search_match:
        return f"CWE-{int(search_match.group(1))}"
    if text.isdigit():
        return f"CWE-{int(text)}"
    return text


def normalize_cwe_ids(
    raw_cwe_ids: list[str],
    cwe_database: CweDatabase,
) -> list[str]:
    """Normalize, validate, and de-duplicate CWE IDs while preserving order."""
    normalized_ids: list[str] = []
    seen_ids: set[str] = set()
    for raw_cwe_id in raw_cwe_ids:
        normalized_id = normalize_cwe_id(raw_cwe_id)
        if cwe_database.get_by_id(normalized_id) is None:
            raise CweIdError(f"Unknown CWE ID: {normalized_id}")
        if normalized_id not in seen_ids:
            normalized_ids.append(normalized_id)
            seen_ids.add(normalized_id)
    return normalized_ids


def parse_cwe_id_list(raw_value: str | None) -> list[str]:
    """Parse a comma-separated CLI CWE list into raw ID tokens."""
    if raw_value is None:
        return []
    return [part.strip() for part in raw_value.split(",") if part.strip()]
